apps/vaporgui/ImageSubtabs.h

```diff
@@ -40,6 +42,49 @@ class ImageVariablesSubtab : public QWidget {
     }
 };
 
+//
+// PWidget for selecting TMS file level-of-detail
+//
+
+class PTMSLODInput : public PLineItem {
+    Q_OBJECT
+    VComboBox* _vComboBox;
+public:
+    PTMSLODInput() : PLineItem("", "TMS level of detail", _vComboBox = new VComboBox({"0"})) {
+        connect( _vComboBox, &VComboBox::IndexChanged, this, &PTMSLODInput::dropdownIndexChanged );
+    }
+
+protected:
+    virtual void updateGUI() const override {
+        ImageParams* rp = dynamic_cast<ImageParams*>( getParams() );
+        VAssert( rp && "Params must be ImageParams" );
+
+        std::string imageFile = rp->GetImagePath();
+        if ( GeoImageTMS::IsTMSFile( imageFile ) ) {
+            _vComboBox->setEnabled( true );
+        }
+        else {
+            _vComboBox->setEnabled( false );             // Disable if not using a TMS image
+            return;
+        }
+
+        std::vector< std::string > options{"default"};
+        int lods = GeoImageTMS::GetNumTMSLODs( imageFile );
```

Comment:
GetNumTMSLODs does multiple file IO operations recursively. We may not want to do this every update.

Reply:
Agreed this is a problem. File system metadata performance on large parallel filesystems (e.g. glade) is awful. A suggested fix would be to only call GetNumTMSLODs() when imageFile changes.

Reply:
I tried caching the file name in this function (updateGUI()) but this function is const, and we can't update a private member here.

Instead, I've written a new param in ImageParams that holds the number of LODs available.  This only gets updated when the image file changes.

Reply:
Great! You are setting the number of LODs when the image TMS file is set. This needs to be wrapped in a Params group.

Reply:
Good catch.  done.

Reply:
> I tried caching the file name in this function (updateGUI()) but this function is const, and we can't update a private member here.
> 
> Instead, I've written a new param in ImageParams that holds the number of LODs available. This only gets updated when the image file changes.

Rather than introduce a new parameter, and a getter that will return the wrong answer if it is called before SetImagePath, why not make use of the C++ 'mutable' keyword inside of updateGUI to allow an exception to the const declaration? This is what the mutable keyword is for, and the model that we use elsewhere for cached (expensive to compute) values returned by const getter functions. See, for example, the use of _minuCache inside of the Grid class.

Reply:
@clyne If you have two image renderers, it will have to keep track of which renderer is currently active and manually keep its cache in sync.

Reply:
Good point.

Reply:
@StasJ , on second thought, why would you have to keep track of the active renderer? If you cache the file name and # LODs then when the active renderer changes the information will still be valid if the current and previous renderers both use same image file. If the two render instances use different image files that will trigger a cache update, but switching renderers doesn't happen at a high enough frequency to be a performance problem.